tools/swift-inspect/Sources/SwiftInspectLinux/PTrace.swift

```diff
@@ -0,0 +1,151 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import LinuxSystemHeaders
+
+public class PTrace {
+  enum PTraceError: Error {
+    case ptraceFailure(_ command: Int32, pid: pid_t, errno: Int32 = get_errno())
+    case waitFailure(pid: pid_t, errno: Int32 = get_errno())
+    case unexpectedWaitStatus(pid: pid_t, status: Int32, sigInfo: siginfo_t? = nil)
+  }
+
+  let pid: pid_t
+
+  // Initializing a PTrace instance attaches to the target process, waits for
+  // it to stop, and leaves it in a stopped state. The caller may resume the
+  // process by calling cont().
+  public init(process pid: pid_t) throws {
+    guard ptrace_attach(pid) != -1 else {
+      throw PTraceError.ptraceFailure(PTRACE_ATTACH, pid: pid)
+    }
+
+    while true {
+      var status: Int32 = 0
+      let result = waitpid(pid, &status, 0)
+      guard result != -1 else {
+        if get_errno() == EINTR { continue }
+        throw PTraceError.waitFailure(pid: pid)
+      }
+
+      precondition(pid == result, "waitpid returned unexpected value \(result)")
+
+      if wIfStopped(status) { break }
+    }
+
+    self.pid = pid
+  }
+
+  deinit { _ = ptrace_detach(self.pid) }
```

Comment:
This seems less than ideal. The deinitializer is not like C++'s destructor and the time of execution is not guaranteed.

Reply:
Good to know. Sounds more like a finalizer? I assume the same is NOT true for `defer` blocks, right? Assuming so, I will add an explicit detach method and update the call site to call it from a `defer` block.

Reply:
`defer` blocks run at the end of the current scope.

`deinit` isn't quite as unpredictable as finalizers in GC'd languages; Swift uses reference counting, so it will generally run the `deinit` as soon as there are no further references to the object, which in _many_ cases is predictable if you have knowledge of the program you're working on. The complication comes if you pass a reference to your object to something else that happens to hold on to a reference, or (worse) sticks your object into an autorelease pool.